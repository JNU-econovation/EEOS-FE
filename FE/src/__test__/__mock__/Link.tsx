export const Link = jest.fn(({ children, href }) => (
  <a role="link" href={href} onClick={(e) => e.preventDefault()}>
    {children}
  </a>
));
