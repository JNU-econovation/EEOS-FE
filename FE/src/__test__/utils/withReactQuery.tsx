import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

const queryClient = new QueryClient();

const withReactQuery = (Component: React.ReactNode) => {
  return () => (
    <QueryClientProvider client={queryClient}>{Component}</QueryClientProvider>
  );
};

export default withReactQuery;
