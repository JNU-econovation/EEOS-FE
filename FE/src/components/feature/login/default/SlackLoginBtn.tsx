import { useSearchParams } from "next/navigation";
import { useEffect } from "react";
import StyledLinkButton from "../../../common/Button/StyledLinkButton";
import { useSlackLoginMutation } from "@/hooks/query/useAuthQuery";

const SlackLoginBtn = () => {
  const clientId = process.env.NEXT_PUBLIC_SLACK_CLIENT_ID;
  const redirectUri = process.env.NEXT_PUBLIC_SLACK_REDIRECT_URI;
  const teamId = process.env.NEXT_PUBLIC_SLACK_TEAM_ID;

  const slackLoginUrl = `https://slack.com/oauth/v2/authorize?client_id=${clientId}&team_id=${teamId}&scope=&user_scope=users.profile:read&redirect_uri=${redirectUri}&prompt=none`;

  const searchParams = useSearchParams();
  const code = searchParams.get("code");

  const { mutate: loginSlack } = useSlackLoginMutation();

  useEffect(() => {
    if (code) {
      loginSlack({ code, redirect_uri: redirectUri });
    }
  }, [code]);

  return (
    <StyledLinkButton
      linkUrl={slackLoginUrl}
      buttonText="슬랙으로 로그인"
      imageUrl="/icons/slack.svg"
      color="slack"
      alt="슬랙 로그인 버튼"
    />
  );
};

export default SlackLoginBtn;
