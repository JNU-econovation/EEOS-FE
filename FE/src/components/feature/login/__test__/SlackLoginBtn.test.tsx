/**
 * @jest-environment jsdom
 */

import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import renderer from "react-test-renderer";
import SlackLoginBtn from "../default/SlackLoginBtn";
import { Link } from "@/__test__/__mock__/Link";
import withReactQuery from "@/__test__/utils/withReactQuery";

jest.mock("next/navigation", () => ({
  useSearchParams: jest.fn(() => ({
    get: jest.fn(() => "code"),
  })),
  useRouter: jest.fn(() => ({
    back: jest.fn(),
    replace: jest.fn(),
  })),
}));

jest.mock("@/hooks/query/useAuthQuery", () => ({
  useSlackLoginMutation: jest.fn(() => ({
    mutate: jest.fn(),
  })),
}));

const MockNextRouterComponent = Link;

jest.mock("next/link", () => {
  return ({ children, href }) => {
    return (
      <MockNextRouterComponent href={href}>{children}</MockNextRouterComponent>
    );
  };
});

describe("SlackLoginBtn", () => {
  let MockNextRouterComponent;

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("올바르게 렌더링 된다", async () => {
    const WrappedComponent = withReactQuery(<SlackLoginBtn />);
    const tree = renderer.create(<WrappedComponent />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("버튼 클릭시 슬랙 로그인 페이지로 이동한다", async () => {
    const user = userEvent.setup();

    render(<SlackLoginBtn />);

    const button = screen.getByRole("link");
    await user.click(button);

    expect(MockNextRouterComponent).toHaveBeenCalled;
  });
});
