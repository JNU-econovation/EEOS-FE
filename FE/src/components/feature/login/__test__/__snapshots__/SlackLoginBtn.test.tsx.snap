// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlackLoginBtn 올바르게 렌더링 된다 1`] = `
<a
  href="https://slack.com/oauth/v2/authorize?client_id=5775976606341.6306752698400&team_id=undefined&scope=&user_scope=users.profile:read&redirect_uri=https://localhost:3000/login/logging-in"
  onClick={[Function]}
  role="link"
>
  <img
    alt="슬랙 로그인 버튼"
    data-nimg="1"
    decoding="async"
    height={24}
    loading="lazy"
    onError={[Function]}
    onLoad={[Function]}
    src="/icons/slack.svg"
    style={
      {
        "color": "transparent",
      }
    }
    width={24}
  />
  <p
    className="text-center font-semibold"
  >
    슬랙으로 로그인
  </p>
</a>
`;
